tsconfig.json

```diff
@@ -0,0 +1,26 @@
+{
+	"compilerOptions": {
+		// Target latest version of ECMAScript.
+		"target": "esnext",
+		// Search under node_modules for non-relative imports.
+		"moduleResolution": "node",
+		// Process & infer types from .js files.
+		"allowJs": true,
+		"jsx": "react",
```

Comment:
If we start compiling the `.tsx` files with the real TypeScript compiler, which I see as the next step, how do we want our pipeline to look like?
1. `tsx` directly to `js` that will be bundled with webpack, with no intermediate Babel step. Then we need the `jsx: react` option and `target: esnext` will be different for evergreen/ie11 builds.
2. `tsx` to `esnext` (optionally with `jsx`), will be further transformed by Babel before bundling.

As long as we have custom Babel transforms (nonstandard JSX for `wp.element`, inline Redux action imports, `AsyncLoad` transform, ...), we need the intermediate step.

Reply:
> the next step, how do we want our pipeline to look like?

This used to be the source of my biggest questions @jsnajdr - that's one reason why I never felt confident in pushing #15390. Compiling with `tsc` provides a number of significant perks (browser compatibility, speed, a native way of robustly handling packages in a monorepo) but leaves a few unanswered questions.

For one I think there are two dominant TypeScript loaders and I'm not clear on the tradeoffs. Maybe we just pick one of them and we'd be fine.

Second, those babel transforms are tricky. Honestly though I think we could consider removing the action type inlining. If I remember correctly that shaved 10 KB off of the bundle sizes, but I would probably be willing to give that up if it let us use TypeScript as a real compiler. Who knows, it might be possible that we could gain that 10 KB back without too much hassle through better code generation.

Concerning `wp.element` I'm wondering how big of a problem that is. Could we find a simple way to exclude those files if indeed it became a problem? I don't know the answer here.

---

If we could eliminate our custom transforms, which seems like a reasonable possibility (doesn't `AsyncLoad` provide functionality which became standard with `import()`?) then I think a great end-goal is eliminating Babel. I don't see it being practical at this moment but maybe we could envision a time when it would be. (`tsc` has no problem compiling in `.js` files)

Reply:
> Concerning wp.element I'm wondering how big of a problem that is. Could we find a simple way to exclude those files if indeed it became a problem? I don't know the answer here.

As far as I know, we don't have any builds that require a mix of `React.createElement` and `wp.element.createElement`, so we might be able to just make sure that whatever build is running is correctly configured.

Reply:
> doesn't `AsyncLoad` provide functionality which became standard with `import()`?)

`AsyncLoad` is certainly ripe for a major update. When it was created, the ESM dynamic `import()` didn't exist, we used CommonJS modules, and `React.lazy` wasn't a built-in React feature. It's entirely possible that after the update, the custom transform won't be needed any more.

Reply:
> Honestly though I think we could consider removing the action type inlining.

Some time ago I had an idea that it could be a webpack plugin instead. The plugin would detect that a module export is a string/numeric constant and would inline it at the import site. The module concatenation and the `sideEffect: false` plugins are conceptually similar: they also analyze the import/export graph and eliminate indirection.

That would make it a linker-time optimization instead of a compiler-time one.

Reply:
> For one I think there are two dominant TypeScript loaders and I'm not clear on the tradeoffs.

Webpack and TypeScript docs describe `ts-loader` as the first choice. The other one is `awesome-typescript-loader`, is that right? They mention good Babel integration as their main advantage over `ts-loader`. And that sounds like something that we might like 🙂

Reply:
> As far as I know, we don't have any builds that require a mix of `React.createElement` and `wp.element.createElement`

This can change quickly as soon as we start importing anything from `@wordpress/components`. Better be prepared 🙂 Not a blocker right now, just let's not create something that will be a blocker in future.

Reply:
> This can change quickly as soon as we start importing anything from @wordpress/components

If we import from there, the code will probably already be transpiled down to wp.element calls.